### PR TITLE
Complete CRUD Routes

### PR DESCRIPTION
**Pull request by ShaneSutro**
- [X] Added CRUD routes in database file
- [X] Added CRUD routes in server index
- [X] De-coupled express app from function to run server
- [X] De-coupled database connecting with schema
- [X] Tests for all CRUD routes
- [X] Lots of linting error fixes

### CODE REVIEW

File: __test__/crud.test.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/* eslint-disable no-undef */
+const request = require('supertest');
+const app = require('../server/app');
+const { SyllabusModel, connect, disconnect } = require('./db.mocks.js');
```
Comment (phucciTr) [inline __test__/crud.test.js:4]: This is the mocked database that's connected to **localhost:27017/test** right?

File: __test__/crud.test.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+describe('Test the root path', () => {
+  test('It should respond to the GET method', async () => {
+    const response = await request(app).get('/');
+    expect(response.statusCode).toBe(200);
+  });
+
+  test('It should allow the insertion of a document', async () => {
+    let records = await SyllabusModel.find();
+    expect(records.length).toBe(0);
+    const response = await request(app).post('/api/syllabus', { body: fixtures.sampleSyllabusString });
```
Comment (phucciTr) [inline __test__/crud.test.js:28]: Suppose this is where the mocked db is getting tested right?

I'm wondering how can the "**request**" from **supertest** module know to to write data to the **localhost:27017/test** instead of the "real db" from the exposed env variables that is **localhost:27017/syllabuses**? Was there any config that you had to set up to make this happen?

I'm asking because when I looked in your **server/index.js** that has all the CRUD routes from **app.js**, that server was connected to **localhost:27017/syllabuses** that was imported from your **db/index.js**, and now that **server/app.js** is imported here for testing. 

I might be having some wrong assumptions though so please correct me

File: __test__/db.mocks.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+const mongoose = require('mongoose');
+const syllabusesSchema = require('../db/data/syllabusesSchema');
+
+const connect = () => new Promise((resolve, reject) => {
+  mongoose.connect('mongodb://localhost:27017/test', { useNewUrlParser: true, useUnifiedTopology: true });
```
Comment (phucciTr) [inline __test__/db.mocks.js:5]: I can see that this is the mocked db used for testing right?

File: db/data/syllabusesModel.js [hunk 4]
```diff
@@ -8,7 +8,10 @@ const MONGO_HOSTNAME = process.env.MONGO_HOSTNAME || 'localhost';
 const MONGO_PORT = process.env.MONGO_PORT || 27017;
 const MONGO_DB = process.env.MONGO_DB || 'syllabuses';
 
-mongoose.connect(`mongodb://${MONGO_HOSTNAME}:${MONGO_PORT}/${MONGO_DB}`, { useNewUrlParser: true, useUnifiedTopology: true });
+const connect = () => new Promise((resolve, reject) => {
+  mongoose.connect(`mongodb://${MONGO_HOSTNAME}:${MONGO_PORT}/${MONGO_DB}`, { useNewUrlParser: true, useUnifiedTopology: true });
```
Comment (phucciTr) [inline db/data/syllabusesModel.js:12]: I'm wondering if you have already populated and started to use this particular db/collection? I

'm asking because the env vars are exposed on github (might not matter at this point for you), and I'm guessing the connection string connects to **localhost**, and to a db named **syllabuses**?

File: db/data/syllabusesModel.js [hunk 5]
```diff
@@ -18,4 +21,7 @@ db.once('open', () => {
 
 const SyllabusModel = mongoose.model('syllabuses', syllabusesSchema);
 
-module.exports = SyllabusModel;
+module.exports = {
+  SyllabusModel,
```
Comment (phucciTr) [inline db/data/syllabusesModel.js:25]: I'm guessing this is exported to your **db/index.js**? And then the **db/index.js** will be exported to your **server/index.js** to use for CRUD ops?

File: db/index.js [hunk 6]
```diff
@@ -27,5 +26,36 @@ const syllabus = (courseNumber, cb) => {
...
+    return response;
+  },
+
+  delete: async (id) => {
+    const response = await SyllabusModel.deleteOne({ id });
+    return response;
+  },
+};
+
+module.exports = {
```
Comment (phucciTr) [inline db/index.js:57]: This will be exported to your **server/app.js** to use for CRUD ops right?

File: server/app.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+const express = require('express');
+const bodyParser = require('body-parser');
+const path = require('path');
+const cors = require('cors');
+const db = require('../db/index.js');
```
Comment (phucciTr) [inline server/app.js:5]: This db is the one that is connected to localhost:27017/syllabuses right? And it is used in your CRUD endpoints below yea?

**Review by phucciTr — COMMENTED** [review]
Everything looks good to go Shane. The only thing that I'm curious about is which db is actually getting tested with the "supertest" module?

File: __test__/crud.test.js (hunk 1 above)
Comment (ShaneSutro) [reply __test__/crud.test.js:4]: @phucciTr Yep, exactly!

File: __test__/crud.test.js (hunk 2 above)
Comment (ShaneSutro) [reply __test__/crud.test.js:28]: @phucciTr No you've got it! One of the interesting things about Mongo is that as long as _some_ mongo database is connected, all of the calls will use whatever the currently-connected instance is - including the database calls in `app.js`.
In production, the database comes from the `db` directory, where the actual mongo connection is set up and connected. This is what I needed to refactor last night before I could test since they needed to be dissociated. If you look in `server/index.js`, you can see that I'm calling the `connect()` function from the live database and then `app.listen()`. For testing, I'm using the mock version of the `connect()` method, which is listening on `localhost:27017/test`, and instead of `app.listen()`, I'm using `supertest` to mock the server functions based on what's in the `server/app.js` file.
In short, for production it's
- `connect()` to production database
- `app.listen()

For testing it's
- mock `connect()` to test database
- `request(app)` (`supertest` *instead* of `app.listen()`)

File: db/data/syllabusesModel.js (hunk 4 above)
Comment (ShaneSutro) [reply db/data/syllabusesModel.js:12]: @phucciTr - You're right, yeah. This is legacy code, I think the original author did it this way to be able to make quick switches between development and production. In my case, I'm not using a `.env` file anyway, so these env variables aren't being used right now.

File: db/data/syllabusesModel.js (hunk 5 above)
Comment (ShaneSutro) [reply db/data/syllabusesModel.js:25]: Exactly! This is also legacy code other than the fact that I needed to also export the `connect()` function, so I had re-write this to export two things instead of just one.

File: server/app.js (hunk 7 above)
Comment (ShaneSutro) [reply server/app.js:5]: This `db` is the production version, yes, kind of - only difference is that I'm not using it's `connect()` function (which would connect it to prod database), so it's still connecting to the test db, not the production `/syllabuses` database. In this case, `db.rest` contains all the CRUD methods, so `db.rest.get()`, `db.rest.create()`, `db.rest.delete()`, etc.

File: __test__/db.mocks.js (hunk 3 above)
Comment (ShaneSutro) [reply __test__/db.mocks.js:5]: @phucciTr So this is actually using the same syllabus schema as `prod`, but yes importing it here allows for easier testing - I'm using the same schema, just a mocked/test version of the `connect()` function found in `db/index.js`.
